server/routes/calculationQuestionRoutes.ts

```diff
@@ -201,6 +206,35 @@ export default class CalculationQuestionRoutes {
     })
   }
 
+  public selectCalculationReason: RequestHandler = async (req, res): Promise<void> => {
+    const { username, caseloads, token } = res.locals.user
+    const { nomsId } = req.params
+    const calculationReasons = await this.calculateReleaseDatesService.getCalculationReasons(res.locals.user.token)
+    const prisonerDetail = await this.prisonerService.getPrisonerDetail(username, nomsId, caseloads, token)
+
+    return res.render('pages/calculation/reason', { reasons: calculationReasons, prisonerDetail })
+  }
+
+  public submitCalculationReason: RequestHandler = async (req, res): Promise<void> => {
+    const { username, caseloads, token } = res.locals.user
+    const { nomsId } = req.params
+    const calculationReasons = await this.calculateReleaseDatesService.getCalculationReasons(res.locals.user.token)
+    const prisonerDetail = await this.prisonerService.getPrisonerDetail(username, nomsId, caseloads, token)
+    const otherId = calculationReasons.find(calculation => calculation.isOther).id
+
+    if (+req.body.calculationReasonId === otherId && req.body.otherReasonDescription.length === 0) {
+      return res.render('pages/calculation/reason', {
+        reasons: calculationReasons,
+        prisonerDetail,
+        otherErrorMessage: { text: 'The reason must not be empty.', id: otherId },
+      })
+    }
+
+    req.session.calculationReasonId = req.body.calculationReasonId
```

Comment:
From memory, and I think there will be other examples of this, its better to associate the reason with the prisoner id, ie
```req.session.calculationReasonId['A1234AA'] = req.body.calculationReasonId```
As users calculate new prisonerIds it means you're not accidentally using the previous prisoners reason for this session if a user goes back and forth through the journey.